Proposal for RemoteWrite CRD (#6781)
prometheus-operator/prometheus-operator: Documentation/proposals/202407-remote-write.md

# RemoteWrite CRD

- Owners:
  - [@superbrothers](https://github.com/superbrothers)
- Related Tickets:
  - [#6508](https://github.com/prometheus-operator/prometheus-operator/issues/6508)
- Other docs:
  - n/a

## TL;DR

This design doc proposes RemoteWrite CRD, which enables cluster admins to delegate the ability to configure Prometheus remote_rewrite configuration to application developers/operators.

## Why

The Prometheus remote_write configuration is defined in the Prometheus CRD. Currently, the configuration data generation is the responsibility of cluster admins.

## Goals

> The main goal is for application developers/operators to configure themselves how the remote endpoint sends metrics.

Reviewer: ```suggestion
The main goal is to enable application developers/operators to self-service the remote write, and configure how the client sends metrics to the remote endpoint.
```

This means exposing a new CRD to configure Prometheus remote_write configuration.

## Non-goals

We do not cover "RemoteRead" CRD here. We can still implement it later if needed.

## How

### RemoteWrite CRD

The RemoteWrite CRD represents one of the Prometheus remote_write configuration scoped to the resource’s namespace.

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: RemoteWrite
metadata:
  name: example
  namespace: default
spec:
  url: "https://aps-workspaces.us-west-2.amazonaws.com/workspaces/<workspace id>/api/v1/remote_write"
  sigv4:
    region: us-west-2

>     accessKey:

Reviewer 2: It might be good to mention that secret/configmap refs will be resolved in the RemoteWrite namespace.

Author: ```suggestion
    # Secret object references will be resolved in the namespace of this remoteWrite object.
    # In this manifest it is the "default" namespace.
    accessKey:
```

      name: aws_access
      key: access_key
    secretKey:
      name: aws_access
      key: secret_key
```

```go
package v1alpha1

import (
	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type RemoteWrite struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec monitoringv1.RemoteWriteSpec `json:"spec"`
}
```

### Prometheus CRD

The Prometheus CRD is extended with 2 new fields (remoteWriteSelector and remoteWriteNamespaceSelector) that define which RemoteWrite resources are associated with this Prometheus instance.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: example
  namespace: default
spec:

>   # RemoteWrite objects to be selected. If nil, nothing will be selected.

Author: Fixed at 4d8afb2c6.

  remoteWriteSelector: {}
  # Namespace(s) to be selected for RemoteWrite discovery. If nil, only check in its own namespace.
  remoteWriteNamespaceSelector: {}
  ...
```

#### Respect enforceNamespaceLabel and excludedFromEnforcement

The RemoteWrite CRD should respect the Prometheus `.spec.enforcedNamespaceLabel` field.

> The object’s namespace is added as the last item in the `write_relabel_config` of the remote_write configuration.

Author: It seems to me that simply replacing the namespace labels does not prevent other tenants from stealing metrics. Should we only keep metrics for the namespace where the RemoteWrite object is created, as follows?

```yaml
write_relabel_configs:
...
- sourceLabels:
  - namespace
  regex: <namespace>
  action: keep
```

> The object’s namespace is added as the last item in the `write_relabel_config` of the remote_write configuration.

Author: Updated in 29a6f2b21ea9c12efaf0ccf65b737b42995dd649.

```yaml
write_relabel_configs:
...
- target_label: <enforced-namespace-label>
  replacement: <namespace>
```

> The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field.

Reviewer 2: Side-note: the new CRD would need to be added to the list of supported resources.

These were proposed at https://github.com/prometheus-operator/prometheus-operator/issues/6508#issuecomment-2058942336.

### PrometheusAgent CRD

> TODO

Reviewer 2: From I can see, it should work the same as the Prometheus CRD. Is there anything that would differ?

Reviewer 2: Thanks! Then it's probably simpler to say this at the beginning and that everything which is described for the Prometheus CRD will work the same for PrometheusAgent.

Author: Updated in 29b96b41a6f0ce1a3acd7d4efeac0a58eab22d40.

### Configuration generation

> The Prometheus operator will generate the Prometheus configuration including remote_write configuration from the Prometheus CRD and the RemoteWrite resources matching remoteWriteSelector from the namespace(s) selected by remoteWriteNamespaceSelector for additional remote_rewrite configuration. The generated configuration will be stored in a secret mounted in the Prometheus pod.

Reviewer 2: > The generated configuration will be stored in a secret mounted in the Prometheus pod.

No change to the current way, right?

Author: Yes, it is still the same. The configuration file is gzipped, but essentially unchanged.

Reviewer 2: Right so let's omit the things which won't change.
```suggestion
The Prometheus operator will generate the Prometheus configuration including remote_write configuration from the Prometheus CRD and the RemoteWrite resources matching remoteWriteSelector from the namespace(s) selected by remoteWriteNamespaceSelector for additional remote_rewrite configuration.
```

- Other docs:
  - n/a

## TL;DR

This design doc proposes RemoteWrite CRD, which enables cluster admins to delegate the ability to configure Prometheus remote_rewrite configuration to application developers/operators.

## Why

The Prometheus remote_write configuration is defined in the Prometheus CRD. Currently, the configuration data generation is the responsibility of cluster admins.

## Goals

The main goal is to enable application developers/operators to self-service the remote write, and configure how the client sends metrics to the remote endpoint.

This means exposing a new CRD to configure Prometheus remote_write configuration.

## Non-goals

We do not cover "RemoteRead" CRD here. We can still implement it later if needed.

## How

### RemoteWrite CRD

The RemoteWrite CRD represents one of the Prometheus remote_write configuration scoped to the resource’s namespace.

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: RemoteWrite
metadata:
  name: example
  namespace: default
spec:
  url: "https://aps-workspaces.us-west-2.amazonaws.com/workspaces/<workspace id>/api/v1/remote_write"
  sigv4:
    region: us-west-2

>     # Secret object references will be resolved in the namespace of this remoteWrite object.
>     # In this manifest it is the "default" namespace.
>     accessKey:
>       name: aws_access
>       key: access_key

Author: ```suggestion
    accessKey:
      # Secret object references will be resolved in the namespace of this remoteWrite object.
      # In this manifest it is the "default" namespace.
      name: aws_access
      key: access_key
```

}
```

### Prometheus CRD

The Prometheus CRD is extended with 2 new fields (remoteWriteSelector and remoteWriteNamespaceSelector) that define which RemoteWrite resources are associated with this Prometheus instance.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: example
  namespace: default
spec:
  # RemoteWrite objects to be selected. An empty label selector matches all
  # objects. A null label selector matches no objects.
  remoteWriteSelector: {}
  # Namespaces to match for RemoteWrite discovery. An empty label selector
  # matches all namespaces. A null label selector matches the current namespace
  # only.
  remoteWriteNamespaceSelector: {}
  ...
```

#### Respect enforceNamespaceLabel and excludedFromEnforcement

The RemoteWrite CRD should respect the Prometheus `.spec.enforcedNamespaceLabel` field.

The object’s namespace is added as the last item in the `write_relabel_config` of the remote_write configuration.

```yaml
write_relabel_configs:
...
- target_label: <enforced-namespace-label>
  replacement: <namespace>
```

The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field.

> 
> These were proposed at https://github.com/prometheus-operator/prometheus-operator/issues/6508#issuecomment-2058942336.

Author: ```suggestion
```

# RemoteWrite CRD

- Owners:
  - [@superbrothers](https://github.com/superbrothers)
- Related Tickets:
  - [#6508](https://github.com/prometheus-operator/prometheus-operator/issues/6508)
- Other docs:
  - n/a

## TL;DR

This design doc proposes RemoteWrite CRD, which enables cluster admins to delegate the ability to configure Prometheus remote_rewrite configuration to application developers/operators.

## Why

The Prometheus remote_write configuration is defined in the Prometheus CRD. Currently, the configuration data generation is the responsibility of cluster admins.

## Goals

The main goal is to enable application developers/operators to self-service the remote write, and configure how the client sends metrics to the remote endpoint.

This means exposing a new CRD to configure Prometheus remote_write configuration.

## Non-goals

> We do not cover "RemoteRead" CRD here. We can still implement it later if needed.

Author: ```suggestion
Prometheus has two APIs for integrating with remote storage: 'remote write' and 'remote read'. Since there has been no confirmed demand to treat remote read as a CRD, we do not cover "RemoteRead" CRD here. We can still implement it later if needed.
```

```

```go
package v1alpha1

import (
	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type RemoteWrite struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec monitoringv1.RemoteWriteSpec `json:"spec"`
}
```

### Prometheus CRD

The Prometheus CRD is extended with 2 new fields (remoteWriteSelector and remoteWriteNamespaceSelector) that define which RemoteWrite resources are associated with this Prometheus instance.

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: example
  namespace: default
spec:
  # RemoteWrite objects to be selected. An empty label selector matches all
  # objects. A null label selector matches no objects.
  remoteWriteSelector: {}
  # Namespaces to match for RemoteWrite discovery. An empty label selector
  # matches all namespaces. A null label selector matches the current namespace
  # only.
  remoteWriteNamespaceSelector: {}
  ...
```

#### Respect enforceNamespaceLabel and excludedFromEnforcement

The RemoteWrite CRD should respect the Prometheus `.spec.enforcedNamespaceLabel` field.

> The object’s namespace is added as the first item in the `write_relabel_config` of the remote_write configuration.

Reviewer 3: Nice approach to respect tenants!

```yaml
write_relabel_configs:
- sourceLabels:
  - namespace
  regex: <namespace>
  action: keep
  ...
```

> The item is inserted in the first position by the operator before any other user relabelings then there's no way that users can have access to "foreign" metrics.

Reviewer 4: Pretty neat. But I am still not sure about how reliable this would be.

Reviewer 4: Might be a nit, but for clarity, could we elaborate on what we mean by "foreign" metrics?

Author: ```suggestion
The item is inserted in the first position by the operator before any other user relabelings, so there's no way that users can have access to metrics labeled with namespaces other than the one in which the RemoteWrite object is created.
```

> The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field.

Reviewer 3: This might sound as a silly question coming from a maintainer, but what does `excludedFromEnforcement` do again? 😅

Reviewer 2: right now it's used to enforce a label (typically "namespace") with the "namespace of origin" value and it's supported for:
* metrics scraped from ServiceMonitor, PodMonitor, Probe objects
* rules from PrometheusRule objects

Author: ```suggestion
The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field. It is a list of references to objects to be excluded from `enforcedNamespaceLabel`.
```

### Configuration generation

The Prometheus operator will generate the Prometheus configuration including remote_write configuration from the Prometheus CRD and the RemoteWrite resources matching remoteWriteSelector from the namespace(s) selected by remoteWriteNamespaceSelector for additional remote_rewrite configuration.

The operator will respect the --namespaces and --deny-namespaces flags when looking for RemoteWrite objects.

> 

Author: Thanks!

The RemoteWrite CRD should respect the Prometheus `.spec.enforcedNamespaceLabel` field.

The object’s namespace is added as the first item in the `write_relabel_config` of the remote_write configuration.

```yaml
write_relabel_configs:
- sourceLabels:
  - namespace
  regex: <namespace>
  action: keep
  ...
```

The item is inserted in the first position by the operator before any other user relabelings, so there's no way that users can have access to metrics labeled with namespaces other than the one in which the RemoteWrite object is created.

The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field. It is a list of references to objects to be excluded from `enforcedNamespaceLabel`.

### Configuration generation

The Prometheus operator will generate the Prometheus configuration including remote_write configuration from the Prometheus CRD and the RemoteWrite resources matching remoteWriteSelector from the namespace(s) selected by remoteWriteNamespaceSelector for additional remote_rewrite configuration.

The operator will respect the --namespaces and --deny-namespaces flags when looking for RemoteWrite objects.

### Administrative limits

> Similar to what exists for scrape resources (e.g. `enforcedSampleLimit` for  `ServiceMonitor`), we will allow the Prometheus resource's owners to setup upper-bound limits on the remote write options. For example, it will be possible to disable metadata sending or limit the queue capacity:

Author: ```suggestion
Similar to what exists for scrape resources (e.g. `enforcedSampleLimit` for `ServiceMonitor`), we will allow the Prometheus resource's owners to setup upper-bound limits on the remote write options. For example, it will be possible to disable metadata sending or limit the queue capacity:
```
